### DISCUSSION TITLE
Using Redux in both host app and mini app

### DISCUSSION THREAD

**Discussion by Akashdeep312** [callstack/repack#735] [Q&A]
I have a mini app which has  full functionality  and it's own store using redux and redux toolkit , now i want to load into host app which is based of [Super App Showcase](https://github.com/callstack/super-app-showcase) and i'm having the following error

![image](https://github.com/user-attachments/assets/622c6f53-2df3-4f9f-aada-56927d03264e)

mini app that i want to load has following configuration.

```js
<Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <PaperProvider>
          <ThemeProvider>
            <GestureHandlerRootView style={{flex: 1}}>
              <Stack.Navigator screenOptions={{headerShown: false}}>
                <Stack.Screen name={'BOOK'} component={Booking} />
              </Stack.Navigator>
            </GestureHandlerRootView>
          </ThemeProvider>
        </PaperProvider>
      </PersistGate>
    </Provider>
```

**Comment by sahajarora1286** [comment]
I accomplished using Redux in React-Native-Web (Re.Pack) by leveraging redux toolkit's `combineSlices` and `injectReducer` functionality. See: https://redux-toolkit.js.org/api/combineSlices#withlazyloadedslices

I have one true Redux store (the global store) at the root, which is specified as a singleton instance via the ModuleFederation plugin.

Mini apps maintain their own slices and expose a `rootSlice` (combinedSlices) via the ModuleFederation config.

When any component in the `host` tries to download/render a remote component from a mini-app, it also first downloads the exposed `rootSlice` and then `injects` it into the `rootReducer` (Reducer of the root store).

So the only place where I render a Redux `<Provider store={store} >` is in the `host` app. The mini apps don't need this provider.

I have a monorepo setup (managed by Nx) and the global store exposes hooks to allow any app in the monorepo to access global store's state (e.g `useAuth`). 

Mini-apps are able to successfully define their own slices, selectors and actions (all defined in their own projects),

**Reply by Akashdeep312** [reply]
My host app has only one different configuration than my mini app and that's as follows:-
```js
//mini-app
"@callstack/repack": "^4.3.1"
"webpack": "^5.94.0",
  
//Host:
"@callstack/repack": "3.7.0",
"webpack": "5.75.0"

```

Okay if i'm following you then i need to remove the mini app provider and have only one provider at host level, but how can i configure host to only download first the rootslice for a perticular mini app before loading the actual mini app, do you have any reference that you can suggest.

**Reply by sahajarora1286** [reply]
Whenever a component in my `host` needs to render any `remote` component, I make it render a `RemoteComponent` passing it the `appName` and the `componentName`. There's custom logic in this RemoteComponent to `loadRemoteModule` from the appropriate module's deployment URL. 

The `loadRemoteModule.ts` utility and the `RemoteComponent.tsx` component exist in a common `remote-loader-lib` library. Whenever `RemoteComponent` is rendered, it first downloads the `rootSlice` module from the `remote` app and then the actual component from that remote app that the consumer of `RemoteComponent` wants to handle. 
So `RemoteComponent.tsx` downloads the `rootSlice` from the mini remote app and then `injects` it into the `rootReducer` of the global redux store.

**Reply by Akashdeep312** [reply]
Hi @sahajarora1286  thanks for the details , i've updated my component with similar details 

My Component that loads the **mini app** Remote component
```js


const RemoteComponent = React.lazy(() => Federated.importModule('testApp', './App'));

const NewsScreen = () => {
  const [isMiniAppLoaded, setIsMiniAppLoaded] = useState(false);
  useEffect(() => {
    const loadMiniAppSlice = async () => {
      try {
        const {default: miniSlice} = await Federated.importModule(
          'testApp',
          './rootReducer',
        );

        injectReducer('mini', miniSlice);
        setIsMiniAppLoaded(true);
      } catch (error) {
        console.error('Failed to load mini-app slice:', error);
      }
    };

    loadMiniAppSlice();
  }, []);

  return (
    <ErrorBoundary name="NewsScreen">
      <React.Suspense
        fallback={<Placeholder label="News and Articles" icon="newspaper" />}>
        {isMiniAppLoaded && <RemoteComponent />}
      </React.Suspense>
    </ErrorBoundary>
  );
};

export default NewsScreen;

```


Inject Reducer in **host app**
```js
import {configureStore, combineReducers, Reducer} from '@reduxjs/toolkit';
import counterReducer from './counter/counterSlice';

const staticReducers = {
  counter: counterReducer,
};

const store = configureStore({
  reducer: staticReducers,
});

export type RootState = ReturnType<typeof store.getState>;
export default store;

export const injectReducer = (key: string, asyncReducer: Reducer) => {
  if (!store.asyncReducers) {
    store.asyncReducers = {...staticReducers};
  }

  if (!store.asyncReducers[key]) {
    store.asyncReducers[key] = asyncReducer;
    store.replaceReducer(combineReducers(store.asyncReducers));
  }
};


```

<img width="730" alt="Screenshot 2024-09-11 at 10 37 56 AM" src="https://github.com/user-attachments/assets/aa7feb98-0be0-4d69-b388-b770a2498215">


Now when i try to load the remote bundle then i'm getting the issues related to immer and redux , i've already added the immer and redux alias in the host and remote app webapp.config.js files like following:-

```js
         redux: path.resolve('./node_modules/redux/dist/redux.legacy-esm.js'),
        'react-redux': path.resolve(
          './node_modules/react-redux/dist/react-redux.legacy-esm.js',
        ),
        immer: path.resolve('./node_modules/immer/dist/immer.legacy-esm.js'),
        '@reduxjs/toolkit': path.resolve(
          './node_modules/@reduxjs/toolkit/dist/redux-toolkit.legacy-esm.js',
        ),
        'redux-thunk': path.resolve(
          './node_modules/redux-thunk/dist/redux-thunk.legacy-esm.js',
        ),

```

**Comment by sahajarora1286** [comment]
What is the exact error that you see?

**Reply by Akashdeep312** [reply]
I get the following errors in my **Host** app

```js

[ScriptManager] Failed while resolving script locator: {"caller": undefined, "scriptId": "hotel"} {"originalError": [TypeError: undefined is not a function]}


```

![image](https://github.com/user-attachments/assets/03aa8b21-4ce0-42d9-aa5f-0533c6ccd5fc)

**Reply by sahajarora1286** [reply]
Please make sure that when you `alias` `immer` and `redux-toolkit` (..etc), the path you specify actually exists.
I have the following aliases :

``` alias: {
        reselect: path.resolve(
          '../../node_modules/reselect/dist/reselect.legacy-esm.js'
        ),
        redux: path.resolve(
          '../../node_modules/redux/dist/redux.legacy-esm.js'
        ),
        '@reduxjs/toolkit/query/react': path.resolve(
          '../../node_modules/@reduxjs/toolkit/dist/query/react/rtk-query-react.legacy-esm.js'
        ),
        '@reduxjs/toolkit/query': path.resolve(
          '../../node_modules/@reduxjs/toolkit/dist/query/rtk-query.legacy-esm.js'
        ),
        '@reduxjs/toolkit': path.resolve(
          '../../node_modules/@reduxjs/toolkit/dist/redux-toolkit.legacy-esm.js'
        ),
        'redux-thunk': path.resolve(
          '../../node_modules/redux-thunk/dist/redux-thunk.legacy-esm.js'
        ),
        immer: path.resolve(
          '../../node_modules/immer/dist/immer.legacy-esm.js'
        ),
        'react-native-safe-area-context': path.resolve(
          '../../node_modules/react-native-safe-area-context'
        ),
      },
```

Note that I have a mono-repo setup and hence the `../../` in the paths.

Also make sure to add the following in the ModuleFederation config of both host and mini-app:

```  
         reselect: {
            singleton: true,
            requiredVersion: monorepoDeps['reselect'],
            eager: true,
          },
          'react-redux': {
            singleton: true,
            requiredVersion: monorepoDeps['react-redux'],
            eager: true,
          },
          '@reduxjs/toolkit': {
            singleton: true,
            requiredVersion: monorepoDeps['@reduxjs/toolkit'],
            eager: true,
          },
```

Replace `monorepoDeps[...]` with the actual versions of those dependencies from your `package.json`.

Also, please make sure to have a babel-loader rule to transpile these libraries. This is what I have as a rule for these libs:

```
 module: {
      /**
       * This rule will process all React Native related dependencies with Babel.
       * If you have a 3rd-party dependency that you need to transpile, you can add it to the
       * `include` list.
       *
       * You can also enable persistent caching with `cacheDirectory` - please refer to:
       * https://github.com/babel/babel-loader#options
       */
      rules: [
        {
          test: /\.[jt]sx?$/,
          include: [
            /node_modules(.*[/\\])+react\//,
            /node_modules(.*[/\\])+react-native/,
            /node_modules(.*[/\\])+@react-native/,
            /node_modules(.*[/\\])+react-freeze/,
            /node_modules(.*[/\\])+@react-navigation/,
            /node_modules(.*[/\\])+@react-native-community/,
            /node_modules(.*[/\\])+expo/,
            /node_modules(.*[/\\])+pretty-format/,
            /node_modules(.*[/\\])+metro/,
            /node_modules(.*[/\\])+abort-controller/,
            /node_modules(.*[/\\])+@callstack\/repack/,
            /node_modules(.*[/\\])+@callstack[/\\]repack/, // For Windows
            /node_modules(.*[/\\])+immer/,
            /node_modules(.*[/\\])+redux/,
            /node_modules(.*[/\\])+react-redux/,
            /node_modules(.*[/\\])+reselect/,
            /node_modules(.*[/\\])+@reduxjs\/toolkit/,
            /node_modules(.*[/\\])+@reduxjs[/\\]toolkit/, // For Windows
            /node_modules(.*[/\\])+@reduxjs(.*[/\\])+toolkit/,
            /node_modules(.*[/\\])+react-error-boundary/,
          ],
          use: {
            loader: 'babel-loader',
            options: {
              // target: 'esm',
              presets: ['module:metro-react-native-babel-preset'],
              cacheDirectory: true,
            },
          },
        },
       ....
```

**Reply by Akashdeep312** [reply]
@sahajarora1286 still getting the same error

```js

Failed to load mini-app slice: [Error: Unhandled error. ([ScriptManager] Failed while resolving script locator:)] 

```

but after adding 

```js
use: {
            loader: 'babel-loader',
            options: {
              // target: 'esm',
              presets: ['module:metro-react-native-babel-preset'],
              cacheDirectory: true,
            },
          },
```

unable to create webpack bundle

I've also verified the steps all looks good to me
